SU2_CFD/src/output/COutput.cpp

```diff
@@ -0,0 +1,1424 @@
+
+  for (unsigned short iField_Output = 0; iField_Output < VolumeOutput_List.size(); iField_Output++){
+    
+    VolumeOutputField &Field = VolumeOutput_Map[VolumeOutput_List[iField_Output]];
+    
+    /*--- Loop through all fields specified in the config ---*/
+    
+    for (unsigned short iReqField = 0; iReqField < nRequestedVolumeFields; iReqField++){
+      
+      RequestedField = RequestedVolumeFields[iReqField];  
+            
+      if (((RequestedField == Field.OutputGroup) || (RequestedField == VolumeOutput_List[iField_Output])) && (Field.Offset == -1)){
+        Field.Offset = GlobalField_Counter;
+        Variable_Names.push_back(Field.FieldName);
+        GlobalField_Counter++;
+        
+        FoundField[iReqField] = true;
+      }
+    }    
+  }
+  
+  for (unsigned short iReqField = 0; iReqField < nRequestedVolumeFields; iReqField++){
+    if (!FoundField[iReqField]){
+      FieldsToRemove.push_back(RequestedVolumeFields[iReqField]);
+    }
+  }
+  
+  /*--- Remove fields which are not defined --- */    
+  
+  for (unsigned short iReqField = 0; iReqField < FieldsToRemove.size(); iReqField++){
+    if (rank == MASTER_NODE) {
+      if (iReqField == 0){
+        cout << "  Info: Ignoring the following volume output fields/groups:" << endl;
+        cout << "  ";
+      }
+      cout << FieldsToRemove[iReqField];
+      if (iReqField != FieldsToRemove.size()-1){
+        cout << ", ";
+      } else {
+        cout << endl;
+      }
+    }
+    RequestedVolumeFields.erase(std::find(RequestedVolumeFields.begin(), RequestedVolumeFields.end(), FieldsToRemove[iReqField]));
+  }
+  
+  if (rank == MASTER_NODE){
+    cout <<"Volume output fields: ";
+    for (unsigned short iReqField = 0; iReqField < nRequestedVolumeFields; iReqField++){
+      RequestedField = RequestedVolumeFields[iReqField];            
+      cout << RequestedVolumeFields[iReqField];
+      if (iReqField != nRequestedVolumeFields - 1) cout << ", ";
+    }
+    cout << endl;
+  }
+  
+  unsigned long nPoint = 0;
+  
+  if (fem_output){
+    
+    /*--- Create an object of the class CMeshFEM_DG and retrieve the necessary
+   geometrical information for the FEM DG solver. ---*/
+    
+    CMeshFEM_DG *DGGeometry = dynamic_cast<CMeshFEM_DG *>(geometry);
+    
+    unsigned long nVolElemOwned = DGGeometry->GetNVolElemOwned();
+    
+    CVolumeElementFEM *volElem  = DGGeometry->GetVolElem();
+    
+    /*--- Access the solution by looping over the owned volume elements. ---*/
+    
+    for(unsigned long l=0; l<nVolElemOwned; ++l) {
+      
+      for(unsigned short j=0; j<volElem[l].nDOFsSol; ++j) {        
+       
+        nPoint++;
+        
+      }
+    }
+  } else {
+    nPoint = geometry->GetnPoint();
+  }
+  
+  /*--- Now that we know the number of fields, create the local data array to temporarily store the volume output 
+   * before writing it to file ---*/
+   
+  Local_Data.resize(nPoint, std::vector<su2double>(GlobalField_Counter, 0.0));
+  
```

Comment:
This is not a very good way to store the data, have you considered flattening this into a single dimension or at least transposing it?

Reply:
After looking at the parallelization part I understand why it cannot be transposed.
I am going to write a re-usable 2D container for CVariable, we could eventually use that.